horovod/common/operations.cc

```diff
@@ -133,44 +141,75 @@ const Status DUPLICATE_NAME_ERROR = Status::InvalidArgument(
   std::vector<std::shared_ptr<AllgatherOp>> allgather_ops;
   std::vector<std::shared_ptr<BroadcastOp>> broadcast_ops;
 
 #if HAVE_CUDA
 #if HOROVOD_GPU_ALLREDUCE == 'M'
-  allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(new MPI_CUDAAllreduce(&mpi_context, &cuda_context, &state)));
+  allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(
+      new MPI_CUDAAllreduce(&mpi_context, &cuda_context, &state)));
 
 #else
-  #if HAVE_NCCL && HOROVOD_GPU_ALLREDUCE == 'N'
-    allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(
-        new NCCLHierarchicalAllreduce(&nccl_context, &mpi_context, &cuda_context, &state)));
-    allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(
-        new NCCLAllreduce(&nccl_context, &mpi_context, &cuda_context, &state)));
-
-  #elif HAVE_DDL && HOROVOD_GPU_ALLREDUCE == 'D'
-    allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(new DDLAllreduce(&ddl_context, &cuda_context, &state)));
-  #endif
+#if HAVE_NCCL && HOROVOD_GPU_ALLREDUCE == 'N'
+  LOG(INFO)<<"NCCL enabled.\n";
+  allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(
+      new NCCLHierarchicalAllreduce(&nccl_context, &mpi_context,
+          &cuda_context, &state)));
+  allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(
+      new NCCLAllreduce(&nccl_context, &mpi_context, &cuda_context, &state)));
+
+#elif HAVE_DDL && HOROVOD_GPU_ALLREDUCE == 'D'
+  LOG(INFO)<<"DDL enabled.\n";
+  allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(
+      new DDLAllreduce(&ddl_context, &cuda_context, &state)));
+#endif
 
-  allgather_ops.push_back(std::shared_ptr<AllgatherOp>(new MPIHierarchicalAllgather(&mpi_context, &state)));
+  allgather_ops.push_back(std::shared_ptr<AllgatherOp>(
+      new MPIHierarchicalAllgather(&mpi_context, &state)));
+#endif
 #endif
+
+#if HAVE_GLOO
+  if (strcasecmp(state.cpu_operation.c_str(), "gloo")==0) {
+    LOG(INFO) << "GLOO enabled.\n";
+    allreduce_ops.push_back(
+        std::shared_ptr<AllreduceOp>(new GlooAllreduce(&gloo_context, &state)));
+    allgather_ops.push_back(
+        std::shared_ptr<AllgatherOp>(new GlooAllgather(&gloo_context, &state)));
+    broadcast_ops.push_back(
+        std::shared_ptr<BroadcastOp>(new GlooBroadcast(&gloo_context, &state)));
+  }
 #endif
 
 #if HAVE_MLSL
-  allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(new MLSLAllreduce(&mlsl_context, &state)));
-  allgather_ops.push_back(std::shared_ptr<AllgatherOp>(new MLSLAllgather(&mlsl_context, &mpi_context, &state)));
-  broadcast_ops.push_back(std::shared_ptr<BroadcastOp>(new MLSLBroadcast(&mlsl_context, &state)));
-#else
-  // Default operations, always enabled but last to be checked.
-  allreduce_ops.push_back(std::shared_ptr<AllreduceOp>(new MPIAllreduce(&mpi_context, &state)));
-  allgather_ops.push_back(std::shared_ptr<AllgatherOp>(new MPIAllgather(&mpi_context, &state)));
-  broadcast_ops.push_back(std::shared_ptr<BroadcastOp>(new MPIBroadcast(&mpi_context, &state)));
+  if (strcasecmp(state.cpu_operation.c_str(), "mlsl")==0) {
```

Comment:
Another thought - during compilation "admins" should be able to set a default for CPU operations.  If MLSL is present, it would be MLSL.  If HOROVOD_CPU_OPERATIONS=GLOO was specified during **installation**, it would default to GLOO even if MPI is present, etc.

What do you think?

Reply:
Yes. `state.cpu_operation` will be set to a default value based on the compilation, if that value is not overridden by user specification.